File: kotlin-multiplatform/src/Token.kt
package monkeylang

enum class TokenType {
    Illegal,
    EndOfFile,

    // Identifiers + literals
    Identifier,
    Integer,
    StringLiteral,

    // Operators
    Assign,
    Plus,
    Minus,
    Slash,
    Asterisk,

    // Comparison
    Equals,
    NotEquals,
    Bang,
    LessThan,
    GreaterThan,

    // Keywords
    Function,
    Let,
    If,
    Else,
    Return,
    True,
    False,

    // Delimiters
    Semicolon,
    Colon,
    Comma,
    LeftParen,
    RightParen,
    LeftSquirly,
    RightSquirly,
    LeftBracket,
    RightBracket,
}

data class Token(val type: TokenType, val literal: String, val range: DocumentRange)

fun lookupIdentifier(identifier: String): TokenType {
    return when (identifier) {
        "fn" -> TokenType.Function
        "let" -> TokenType.Let
        "if" -> TokenType.If
        "else" -> TokenType.Else
        "return" -> TokenType.Return
        "true" -> TokenType.True
        "false" -> TokenType.False
        else -> TokenType.Identifier
    }
}

File: rust_blaz/lexer/src/token.rs
use std::fmt::Display;

#[derive(Debug, PartialEq)]
pub enum Token {
    Illegal,
    Eof,
    // Identifiers + literals
    Ident(String),
    Int(String),
    // Operators
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,
    LessThan,
    GreaterThan,
    Equal,
    NotEqual,
    // Delimiters
    Comma,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,
    // Keywords
    Function,
    Let,
    True,
    False,
    If,
    Else,
    Return,
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Illegal => write!(f, ""),
            Token::Eof => write!(f, "\0"),
            Token::Ident(value) | Token::Int(value) => write!(f, "{value}"),
            Token::Assign => write!(f, "="),
            Token::Plus => write!(f, "+"),
            Token::Minus => write!(f, "-"),
            Token::Bang => write!(f, "!"),
            Token::Asterisk => write!(f, "*"),
            Token::Slash => write!(f, "/"),
            Token::LessThan => write!(f, "<"),
            Token::GreaterThan => write!(f, ">"),
            Token::Equal => write!(f, "=="),
            Token::NotEqual => write!(f, "!="),
            Token::Comma => write!(f, ","),
            Token::Semicolon => write!(f, ";"),
            Token::LParen => write!(f, "("),
            Token::RParen => write!(f, ")"),
            Token::LBrace => write!(f, "{{"),
            Token::RBrace => write!(f, "}}"),
            Token::Function => write!(f, "fn"),
            Token::Let => write!(f, "let"),
            Token::True => write!(f, "true"),
            Token::False => write!(f, "false"),
            Token::If => write!(f, "if"),
            Token::Else => write!(f, "else"),
            Token::Return => write!(f, "return"),
        }
    }
}

impl From<char> for Token {
    fn from(ch: char) -> Self {
        match ch {
            '=' => Self::Assign,
            ';' => Self::Semicolon,
            '(' => Self::LParen,
            ')' => Self::RParen,
            ',' => Self::Comma,
            '+' => Self::Plus,
            '-' => Self::Minus,
            '!' => Self::Bang,
            '*' => Self::Asterisk,
            '/' => Self::Slash,
            '<' => Self::LessThan,
            '>' => Self::GreaterThan,
            '{' => Self::LBrace,
            '}' => Self::RBrace,
            '\0' => Self::Eof,
            _ => Self::Illegal,
        }
    }
}

impl From<String> for Token {
    fn from(value: String) -> Self {
        match value.as_str() {
            "==" => Self::Equal,
            "!=" => Self::NotEqual,
            "fn" => Self::Function,
            "let" => Self::Let,
            "true" => Self::True,
            "false" => Self::False,
            "if" => Self::If,
            "else" => Self::Else,
            "return" => Self::Return,
            _ => {
                if value.chars().all(|b| b.is_ascii_digit()) {
                    Self::Int(value)
                } else {
                    Self::Ident(value)
                }
            }
        }
    }
}
